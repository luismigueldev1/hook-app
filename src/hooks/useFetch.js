import { useEffect, useRef, useState } from "react";

export function useFetch(url) {
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({ data: null, loading: true, error: null });
    fetch(url)
      .then((response) => response.json())
      .then((data) => {
        isMounted.current &&
          setTimeout(() => {
            setState({
              data: data,
              loading: false,
              error: null,
            });
          }, 4000);
      })
      .catch(() => {
        isMounted.current &&
          setState({
            data: null,
            loading: false,
            error: "No se pudo cargar la info",
          });
      });
  }, [url]);
  return state;
}
