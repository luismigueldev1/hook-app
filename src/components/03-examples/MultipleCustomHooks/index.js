import React from "react";
import { useCounter } from "../../../hooks/useCounter";
import { useFetch } from "../../../hooks/useFetch";

export default function MultipleCustomHooks() {
  const { state: counter, decrement, increment } = useCounter(1);
  const { data, loading, error } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { author, quote } = !!data && data[0];
  //*devuelve los campos de un objeto dentro de un array array
  return (
    <div>
      <h1>Breaking Bad Quotes</h1>
      {error && (
        <div className="alert alert-danger">
          <p>{error.message}</p>
        </div>
      )}

      {loading ? (
        <div className="d-flex justify-content-center">
          <div className="spinner-border text-success " role="status">
            <span className="sr-only">Loading...</span>
          </div>
        </div>
      ) : (
        <blockquote className="blockquote text-right">
          <i> {quote}</i>
          <footer className="blockquote-footer">{author}</footer>
        </blockquote>
      )}

      <div className="text-right mt-2">
        {counter > 1 && (
          <button className="btn btn-secondary" onClick={decrement}>
            Back to previous quote
          </button>
        )}

        <button className="btn btn-primary ml-2" onClick={increment}>
          Get a new quote
        </button>
      </div>
    </div>
  );
}
