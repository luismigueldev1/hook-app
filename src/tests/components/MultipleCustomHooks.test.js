import React from "react";
import { shallow } from "enzyme";
import MultipleCustomHooks from "../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../hooks/useFetch";

jest.mock("../../hooks/useFetch");

describe("Pruebas en <MultipleCustomHooks/>", () => {
  test("debe renderizar el componente", () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("debe mostrar la informaciÃ³n", () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: "Luis Miguel",
          quote: "Hola Mundo",
        },
      ],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    //console.log(wrapper.html());

    expect(wrapper.find(".spinner-border").exists()).toBe(false);
    expect(wrapper.find("i").text().trim()).toBe("Hola Mundo");
    expect(wrapper.find(".blockquote-footer").text().trim()).toBe(
      "Luis Miguel"
    );
  });
});
