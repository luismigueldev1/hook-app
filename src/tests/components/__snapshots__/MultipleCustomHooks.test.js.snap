// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks/> debe renderizar el componente 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <div
    className="d-flex justify-content-center"
  >
    <div
      className="spinner-border text-success "
      role="status"
    >
      <span
        className="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
  <div
    className="text-right mt-2"
  >
    <button
      className="btn btn-primary ml-2"
      onClick={[Function]}
    >
      Get a new quote
    </button>
  </div>
</div>
`;
